tests/check_framework/test_model_field_deprecation.py

```diff
@@ -0,0 +1,83 @@
+from django.db import models
+from django.core import checks
+from django.test import SimpleTestCase
+
+
+class TestDeprecatedField(SimpleTestCase):
+    def test_default_details(self):
+        class DeprecatedField(models.Field):
+            system_check_deprecated_details = {}
+
+        class Model(models.Model):
+            name = DeprecatedField()
+
+        model = Model()
+        self.assertEqual(model.check(), [
+            checks.Warning(
+                msg='DeprecatedField has been deprecated.',
+                hint=None,
+                obj=Model._meta.get_field('name'),
+                id='fields.WXXX',
+            )
+        ])
+
+    def test_user_specified_details(self):
+        class DeprecatedField(models.Field):
+            system_check_deprecated_details = {
+                'msg': 'This field is deprecated and will be removed soon.',
+                'hint': 'Use something else.',
+                'id': 'fields.W999',
+            }
+
+        class Model(models.Model):
+            name = DeprecatedField()
+
+        model = Model()
+        self.assertEqual(model.check(), [
+            checks.Warning(
+                msg='This field is deprecated and will be removed soon.',
+                hint='Use something else.',
+                obj=Model._meta.get_field('name'),
+                id='fields.W999',
+            )
+        ])
+
+
+class TestRemovedField(SimpleTestCase):
+    def test_default_details(self):
+        class DeprecatedField(models.Field):
```

Comment:
I thought about this, but then I thought, "well you don't rename a field when you remove it". Maybe I should just use something generic like `MyField` for both cases.